content/6_Advanced/SCC.mdx

```diff
@@ -99,6 +99,138 @@ int main() {
 
 </CPPSection>
 
+
+<JavaSection>
+
+```java
+
+import java.io.*;
+import java.util.*;
+
+public class Main {
+	static boolean[] vis;
+	static List<Integer>[] adj;    // // Adjacency list of neighbor
+	static List<Integer>[] adj_t;  // adj_t is the transpose of adj
+	static int[] id;
+	static List<Integer> S;
+	public static void main(String[] args) {
+		FastIO io = new FastIO();
+		int n = io.nextInt();
+		int m = io.nextInt();
+		adj = new ArrayList[n + 1];
+		adj_t = new ArrayList[n + 1];
+		vis = new boolean[n + 1];
+		id = new int[n + 1];
+		S = new ArrayList<>();
+
+		for (int i = 1; i <= n; i++) {
+			adj[i] = new ArrayList<>();
+			adj_t[i] = new ArrayList<>();
+		}
+
+		for (int i = 0; i < m; i++) {
+			int a = io.nextInt();
+			int b = io.nextInt();
+			adj[a].add(b);
+			adj_t[b].add(a);
+		}
+
+		for (int i = 1; i <= n; i++) {
+			if (!vis[i]) dfs(i, 1, i);
+		}
+
+		Arrays.fill(vis, false);
+
+		int components = 0;
+		for (int i = n - 1; i >= 0; i--) {
+			if (!vis[S.get(i)]) {
+				components++;
+				dfs(S.get(i), 2, components);
+			}
+		}
+
+		io.println(components);
+		for (int i = 1; i <= n; i++) io.print(id[i] + " ");
+		io.close();
+	}
+	private static void dfs(int x, int pass, int num) {
+		vis[x] = true;
+		List<Integer> ad = (pass == 1) ? adj[x] : adj_t[x];
+		for (int e : ad) {
+			if (!vis[e]) { dfs(e, pass, num); }
+		}
+		S.add(x);
+		if (pass == 2) id[x] = num;
+	}
+
+	// BeginCodeSnip{FastIO}
+	private static class FastIO extends PrintWriter {
```

Comment:
can you use [kattio](https://usaco.guide/general/adding-solution?lang=java#general)

Reply:
Sure.